mm2src/coins/lp_coins.rs

```diff
@@ -1519,6 +1536,33 @@ pub struct ValidateMakerPaymentArgs<'a, Coin: CoinAssocTypes + ?Sized> {
     pub swap_unique_data: &'a [u8],
 }
 
+pub struct ValidateNftMakerPaymentArgs<'a, Coin: CoinAssocTypes + ?Sized> {
+    /// Maker payment tx
+    pub maker_payment_tx: &'a Coin::Tx,
+    /// Maker will be able to refund the payment after this timestamp
+    pub time_lock: u64,
+    /// The hash of the secret generated by taker, this is used for immediate refund
+    pub taker_secret_hash: &'a [u8],
+    /// The hash of the secret generated by maker, taker needs it to spend the payment
+    pub maker_secret_hash: &'a [u8],
+    /// Payment amount
+    pub amount: BigDecimal,
+    /// Taker's HTLC pubkey
+    pub taker_pub: &'a Coin::Pubkey,
+    /// Maker's HTLC pubkey
+    pub maker_pub: &'a Coin::Pubkey,
+    /// Unique data of specific swap
+    pub swap_unique_data: &'a [u8],
+    /// The address of the NFT token
+    pub token_address: &'a [u8],
+    /// The ID of the NFT token.
+    pub token_id: &'a [u8],
+    /// The type of smart contract that governs this NFT
+    pub contract_type: &'a [u8],
+    /// Etomic swap contract address
+    pub swap_contract_address: &'a [u8],
+}
+
```

Comment:
most of the fields here duplicates some in `SendNftMakerPaymentArgs` can we group this field and reuse ? e.g secret_hash can be grouped into a new struct like
```rust
pub struct MakerTakerSecretHash {
  maker: &'a [u8],
  taker: &'a [u8]
}

pub struct MakerTakerPub {
  maker: &'a Coin::Pubkey,
  taker: &'a Coin::Pubkey
}
```

same for `contract_type` and `swap_contract_address` these can then be reused in many places..what do you think?

Reply:
I dont think it is worth creating the structure only for two fields.

```rust
        let secret_hashes = SecretHashes {
            taker_secret_hash: &self.negotiation_data.taker_secret_hash,
            maker_secret_hash: &state_machine.secret_hash(),
        };

        let args = SendMakerPaymentArgs {
            time_lock: state_machine.maker_payment_locktime(),
            secret_hashes: &secret_hashes,
            taker_pub: &self.negotiation_data.maker_coin_htlc_pub_from_taker,
            amount: state_machine.maker_volume.to_decimal(),
            swap_unique_data: &unique_data,
        };
        let maker_payment = match state_machine.maker_coin.send_maker_payment_v2(args).await {
            Ok(tx) => tx,
            Err(e) => {
                let reason = AbortReason::FailedToSendMakerPayment(format!("{:?}", e));
                return Self::change_state(Aborted::new(reason), state_machine).await;
            },
        };
```
for me this is more readable 
```rust
        let args = SendMakerPaymentArgs {
            time_lock: state_machine.maker_payment_locktime(),
            maker_secret_hash: &state_machine.secret_hash(),
            taker_secret_hash: &self.negotiation_data.taker_secret_hash,
            taker_pub: &self.negotiation_data.maker_coin_htlc_pub_from_taker,
            amount: state_machine.maker_volume.to_decimal(),
            swap_unique_data: &unique_data,
        };
        let maker_payment = match state_machine.maker_coin.send_maker_payment_v2(args).await {
            Ok(tx) => tx,
            Err(e) => {
                let reason = AbortReason::FailedToSendMakerPayment(format!("{:?}", e));
                return Self::change_state(Aborted::new(reason), state_machine).await;
            },
        };
``` 

I added this structure into `SendNftMakerPaymentArgs` and `ValidateNftMakerPaymentArgs`
```rust
/// Structure representing necessary NFT info for Swap
pub struct NftSwapInfo<'a> {
    /// The address of the NFT token
    pub token_address: &'a [u8],
    /// The ID of the NFT token.
    pub token_id: &'a [u8],
    /// The type of smart contract that governs this NFT
    pub contract_type: &'a [u8],
    /// Etomic swap contract address
    pub swap_contract_address: &'a [u8],
}
```